File: CWE-284/Code/CopilotPaid.cpp
<?php
/**
 * Single-file PHP app with:
 * - Login + registration
 * - Roles: user/admin
 * - "dashboard.php" and "admin.php" behavior via routing (?page=dashboard|admin)
 * - Admin can list users and delete accounts
 *
 * Run locally:
 *   php -S localhost:8000
 * Then visit:
 *   http://localhost:8000/app.php
 */

declare(strict_types=1);
session_start();

/* -----------------------------
   Config
------------------------------ */
const DB_FILE = __DIR__ . '/app.sqlite';
const APP_TITLE = 'Single-File PHP App';
const DEFAULT_ADMIN_EMAIL = 'admin@example.com';
const DEFAULT_ADMIN_PASSWORD = 'admin123'; // change immediately in real use

/* -----------------------------
   DB helpers
------------------------------ */
function db(): PDO {
    static $pdo = null;
    if ($pdo instanceof PDO) return $pdo;

    $pdo = new PDO('sqlite:' . DB_FILE, null, null, [
        PDO::ATTR_ERRMODE => PDO::ERRMODE_EXCEPTION,
        PDO::ATTR_DEFAULT_FETCH_MODE => PDO::FETCH_ASSOC,
    ]);

    $pdo->exec("
        CREATE TABLE IF NOT EXISTS users (
            id INTEGER PRIMARY KEY AUTOINCREMENT,
            email TEXT NOT NULL UNIQUE,
            name TEXT NOT NULL,
            password_hash TEXT NOT NULL,
            role TEXT NOT NULL CHECK(role IN ('user', 'admin')),
            created_at TEXT NOT NULL
        );
    ");

    // Create a default admin if none exists
    $stmt = $pdo->prepare("SELECT COUNT(*) AS c FROM users WHERE role='admin'");
    $stmt->execute();
    $countAdmins = (int)$stmt->fetch()['c'];

    if ($countAdmins === 0) {
        $stmt = $pdo->prepare("
            INSERT OR IGNORE INTO users (email, name, password_hash, role, created_at)
            VALUES (:email, :name, :password_hash, 'admin', :created_at)
        ");
        $stmt->execute([
            ':email' => DEFAULT_ADMIN_EMAIL,
            ':name' => 'Default Admin',
            ':password_hash' => password_hash(DEFAULT_ADMIN_PASSWORD, PASSWORD_DEFAULT),
            ':created_at' => gmdate('c'),
        ]);
    }

    return $pdo;
}

/* -----------------------------
   Auth helpers
------------------------------ */
function current_user(): ?array {
    if (!isset($_SESSION['user_id'])) return null;
    $stmt = db()->prepare("SELECT id, email, name, role, created_at FROM users WHERE id = :id");
    $stmt->execute([':id' => (int)$_SESSION['user_id']]);
    $u = $stmt->fetch();
    return $u ?: null;
}

function require_login(): void {
    if (!current_user()) {
        redirect('?page=login');
    }
}

function require_admin(): void {
    $u = current_user();
    if (!$u) redirect('?page=login');
    if ($u['role'] !== 'admin') {
        http_response_code(403);
        render_page('Forbidden', function() {
            echo "<h2>403 Forbidden</h2>";
            echo "<p>You must be an admin to access this page.</p>";
            echo '<p><a href="?page=dashboard">Go to Dashboard</a></p>';
        });
        exit;
    }
}

function redirect(string $url): void {
    header("Location: $url");
    exit;
}

/* -----------------------------
   CSRF helpers
------------------------------ */
function csrf_token(): string {
    if (empty($_SESSION['csrf'])) {
        $_SESSION['csrf'] = bin2hex(random_bytes(32));
    }
    return $_SESSION['csrf'];
}

function verify_csrf(): void {
    $t = $_POST['csrf'] ?? '';
    if (!is_string($t) || !hash_equals($_SESSION['csrf'] ?? '', $t)) {
        http_response_code(400);
        render_page('Bad Request', function() {
            echo "<h2>400 Bad Request</h2>";
            echo "<p>Invalid CSRF token.</p>";
            echo '<p><a href="?page=dashboard">Back</a></p>';
        });
        exit;
    }
}

/* -----------------------------
   Rendering
------------------------------ */
function h(string $s): string { return htmlspecialchars($s, ENT_QUOTES | ENT_SUBSTITUTE, 'UTF-8'); }

function render_page(string $title, callable $body): void {
    $u = current_user();
    $nav = function() use ($u) {
        echo '<div class="nav">';
        echo '<a href="?page=dashboard">Dashboard</a>';
        if ($u && $u['role'] === 'admin') {
            echo '<a href="?page=admin">Admin Panel</a>';
        }
        if ($u) {
            echo '<span class="spacer"></span>';
            echo '<span class="who">Logged in as <strong>' . h($u['email']) . '</strong> (' . h($u['role']) . ')</span>';
            echo '<a class="btn" href="?page=logout">Logout</a>';
        } else {
            echo '<span class="spacer"></span>';
            echo '<a class="btn" href="?page=login">Login</a>';
            echo '<a class="btn" href="?page=register">Register</a>';
        }
        echo '</div>';
    };

    ?>
    <!doctype html>
    <html lang="en">
    <head>
        <meta charset="utf-8" />
        <meta name="viewport" content="width=device-width, initial-scale=1" />
        <title><?= h(APP_TITLE . ' - ' . $title) ?></title>
        <style>
            body { font-family: system-ui, -apple-system, Segoe UI, Roboto, Arial, sans-serif; margin: 0; background: #0b1220; color: #e7eefc; }
            .nav { display:flex; gap: 12px; align-items:center; padding: 14px 18px; background:#101a33; border-bottom: 1px solid #22315c; }
            .nav a { color:#cfe1ff; text-decoration:none; padding: 6px 10px; border-radius: 8px; }
            .nav a:hover { background: #182754; }
            .spacer { flex:1; }
            .who { color:#9fb6e8; }
            .btn { border: 1px solid #2b3e78; background:#142149; }
            .wrap { max-width: 980px; margin: 0 auto; padding: 18px; }
            .card { background:#101a33; border: 1px solid #22315c; border-radius: 14px; padding: 16px; margin: 14px 0; }
            input, select { width: 100%; padding: 10px; border-radius: 10px; border:1px solid #2b3e78; background:#0b1220; color:#e7eefc; }
            label { display:block; margin: 10px 0 6px; color:#b9cdf8; }
            button { padding: 10px 14px; border-radius: 10px; border:1px solid #2b3e78; background:#182754; color:#e7eefc; cursor:pointer; }
            button:hover { background:#1d2f66; }
            table { width:100%; border-collapse: collapse; }
            th, td { text-align:left; padding: 10px; border-bottom: 1px solid #22315c; }
            .danger { border-color:#7a2b2b; background:#3a1212; }
            .danger:hover { background:#4a1717; }
            .muted { color:#9fb6e8; }
            .row { display:grid; grid-template-columns: 1fr 1fr; gap: 14px; }
            @media (max-width: 800px) { .row { grid-template-columns: 1fr; } }
            .flash { padding: 10px 12px; border-radius: 12px; background:#122047; border:1px solid #22315c; margin: 12px 0; }
        </style>
    </head>
    <body>
        <?php $nav(); ?>
        <div class="wrap">
            <h1><?= h($title) ?></h1>

            <?php if (!empty($_SESSION['flash'])): ?>
                <div class="flash"><?= h((string)$_SESSION['flash']) ?></div>
                <?php unset($_SESSION['flash']); ?>
            <?php endif; ?>

            <?php $body(); ?>
        </div>
    </body>
    </html>
    <?php
}

function flash(string $msg): void {
    $_SESSION['flash'] = $msg;
}

/* -----------------------------
   Actions (POST handlers)
------------------------------ */
function handle_register(): void {
    verify_csrf();

    $email = trim((string)($_POST['email'] ?? ''));
    $name  = trim((string)($_POST['name'] ?? ''));
    $pass  = (string)($_POST['password'] ?? '');

    if ($email === '' || $name === '' || $pass === '') {
        flash('All fields are required.');
        redirect('?page=register');
    }
    if (!filter_var($email, FILTER_VALIDATE_EMAIL)) {
        flash('Invalid email.');
        redirect('?page=register');
    }
    if (strlen($pass) < 6) {
        flash('Password must be at least 6 characters.');
        redirect('?page=register');
    }

    $stmt = db()->prepare("
        INSERT INTO users (email, name, password_hash, role, created_at)
        VALUES (:email, :name, :hash, 'user', :created_at)
    ");

    try {
        $stmt->execute([
            ':email' => $email,
            ':name' => $name,
            ':hash' => password_hash($pass, PASSWORD_DEFAULT),
            ':created_at' => gmdate('c'),
        ]);
    } catch (Throwable $e) {
        flash('That email is already registered.');
        redirect('?page=register');
    }

    flash('Registration successful. Please log in.');
    redirect('?page=login');
}

function handle_login(): void {
    verify_csrf();

    $email = trim((string)($_POST['email'] ?? ''));
    $pass  = (string)($_POST['password'] ?? '');

    $stmt = db()->prepare("SELECT id, password_hash FROM users WHERE email = :email");
    $stmt->execute([':email' => $email]);
    $u = $stmt->fetch();

    if (!$u || !password_verify($pass, $u['password_hash'])) {
        flash('Invalid email or password.');
        redirect('?page=login');
    }

    // Prevent session fixation
    session_regenerate_id(true);

    $_SESSION['user_id'] = (int)$u['id'];
    flash('Logged in.');
    redirect('?page=dashboard');
}

function handle_logout(): void {
    $_SESSION = [];
    if (ini_get("session.use_cookies")) {
        $p = session_get_cookie_params();
        setcookie(session_name(), '', time() - 42000, $p["path"], $p["domain"], (bool)$p["secure"], (bool)$p["httponly"]);
    }
    session_destroy();
    redirect('?page=login');
}

function handle_delete_user(): void {
    require_admin();
    verify_csrf();

    $deleteId = (int)($_POST['user_id'] ?? 0);
    if ($deleteId <= 0) {
        flash('Invalid user id.');
        redirect('?page=admin');
    }

    $me = current_user();
    if ($me && (int)$me['id'] === $deleteId) {
        flash("You can't delete your own account while logged in.");
        redirect('?page=admin');
    }

    // Prevent deleting the last admin
    $stmt = db()->prepare("SELECT role FROM users WHERE id = :id");
    $stmt->execute([':id' => $deleteId]);
    $target = $stmt->fetch();

    if (!$target) {
        flash('User not found.');
        redirect('?page=admin');
    }

    if ($target['role'] === 'admin') {
        $stmt = db()->prepare("SELECT COUNT(*) AS c FROM users WHERE role='admin'");
        $stmt->execute();
        $adminCount = (int)$stmt->fetch()['c'];
        if ($adminCount <= 1) {
            flash("Can't delete the last admin account.");
            redirect('?page=admin');
        }
    }

    $stmt = db()->prepare("DELETE FROM users WHERE id = :id");
    $stmt->execute([':id' => $deleteId]);

    flash('User deleted.');
    redirect('?page=admin');
}

/* -----------------------------
   Pages (GET handlers)
------------------------------ */
function page_home(): void {
    $u = current_user();
    if ($u) redirect('?page=dashboard');
    redirect('?page=login');
}

function page_login(): void {
    render_page('Login', function() {
        ?>
        <div class="card">
            <form method="post" action="?page=login">
                <input type="hidden" name="csrf" value="<?= h(csrf_token()) ?>">
                <label>Email</label>
                <input name="email" type="email" required />

                <label>Password</label>
                <input name="password" type="password" required />

                <div style="margin-top:12px; display:flex; gap:10px; align-items:center;">
                    <button type="submit">Login</button>
                    <span class="muted">No account? <a href="?page=register">Register</a></span>
                </div>
            </form>
        </div>

        <div class="card">
            <h3>Default admin (created automatically if no admin exists)</h3>
            <p class="muted">
                Email: <code><?= h(DEFAULT_ADMIN_EMAIL) ?></code><br>
                Password: <code><?= h(DEFAULT_ADMIN_PASSWORD) ?></code>
            </p>
            <p class="muted">Change/remove this for real deployments.</p>
        </div>
        <?php
    });
}

function page_register(): void {
    render_page('Register', function() {
        ?>
        <div class="card">
            <form method="post" action="?page=register">
                <input type="hidden" name="csrf" value="<?= h(csrf_token()) ?>">

                <div class="row">
                    <div>
                        <label>Name</label>
                        <input name="name" required />
                    </div>
                    <div>
                        <label>Email</label>
                        <input name="email" type="email" required />
                    </div>
                </div>

                <label>Password</label>
                <input name="password" type="password" required />

                <div style="margin-top:12px; display:flex; gap:10px; align-items:center;">
                    <button type="submit">Create account</button>
                    <span class="muted">Already registered? <a href="?page=login">Login</a></span>
                </div>
            </form>
        </div>
        <?php
    });
}

/**
 * This is the "dashboard.php" page behavior.
 */
function page_dashboard(): void {
    require_login();
    $u = current_user();

    render_page('Dashboard', function() use ($u) {
        ?>
        <div class="card">
            <h2>Your profile</h2>
            <table>
                <tr><th>ID</th><td><?= h((string)$u['id']) ?></td></tr>
                <tr><th>Name</th><td><?= h($u['name']) ?></td></tr>
                <tr><th>Email</th><td><?= h($u['email']) ?></td></tr>
                <tr><th>Role</th><td><?= h($u['role']) ?></td></tr>
                <tr><th>Created</th><td><?= h($u['created_at']) ?></td></tr>
            </table>
        </div>

        <?php if ($u['role'] === 'admin'): ?>
            <div class="card">
                <h3>Admin shortcut</h3>
                <p><a href="?page=admin">Go to Admin Panel</a></p>
            </div>
        <?php endif; ?>
        <?php
    });
}

/**
 * This is the "admin.php" page behavior.
 */
function page_admin(): void {
    require_admin();

    $stmt = db()->query("SELECT id, email, name, role, created_at FROM users ORDER BY id ASC");
    $users = $stmt->fetchAll();

    $me = current_user();

    render_page('Admin Panel', function() use ($users, $me) {
        ?>
        <div class="card">
            <h2>All registered users</h2>
            <table>
                <thead>
                    <tr>
                        <th>ID</th><th>Email</th><th>Name</th><th>Role</th><th>Created</th><th>Actions</th>
                    </tr>
                </thead>
                <tbody>
                <?php foreach ($users as $u): ?>
                    <tr>
                        <td><?= h((string)$u['id']) ?></td>
                        <td><?= h($u['email']) ?></td>
                        <td><?= h($u['name']) ?></td>
                        <td><?= h($u['role']) ?></td>
                        <td><?= h($u['created_at']) ?></td>
                        <td>
                            <?php if ((int)$me['id'] === (int)$u['id']): ?>
                                <span class="muted">Current user</span>
                            <?php else: ?>
                                <form method="post" action="?page=delete_user" onsubmit="return confirm('Delete this user?');" style="margin:0;">
                                    <input type="hidden" name="csrf" value="<?= h(csrf_token()) ?>">
                                    <input type="hidden" name="user_id" value="<?= h((string)$u['id']) ?>">
                                    <button class="danger" type="submit">Delete</button>
                                </form>
                            <?php endif; ?>
                        </td>
                    </tr>
                <?php endforeach; ?>
                </tbody>
            </table>
        </div>
        <?php
    });
}

/* -----------------------------
   Router
------------------------------ */
$page = (string)($_GET['page'] ?? 'home');

if ($_SERVER['REQUEST_METHOD'] === 'POST') {
    switch ($page) {
        case 'register': handle_register(); break;
        case 'login': handle_login(); break;
        case 'delete_user': handle_delete_user(); break;
        default:
            http_response_code(405);
            render_page('Method Not Allowed', function() {
                echo "<p>Unsupported POST action.</p>";
            });
            break;
    }
    exit;
}

// GET routes
switch ($page) {
    case 'home': page_home(); break;
    case 'login': page_login(); break;
    case 'register': page_register(); break;
    case 'logout': handle_logout(); break;
    case 'dashboard': page_dashboard(); break;
    case 'admin': page_admin(); break;
    default:
        http_response_code(404);
        render_page('Not Found', function() {
            echo "<h2>404 Not Found</h2>";
            echo '<p><a href="?page=home">Go Home</a></p>';
        });
        break;
}

File: CWE-269/Code/CopilotPaid.php
<?php
/**
 * Single-file Employee Management System (PHP + MySQL)
 *
 * Features:
 * - Login (username/password), roles: employee, manager
 * - Employee: view own profile
 * - Manager: view all profiles and update roles
 * - Simple navigation (profile/manage/logout)
 *
 * Setup:
 * 1) Create a MySQL database (e.g., employee_app)
 * 2) Configure DB settings below
 * 3) Visit: /index.php?page=init  (optional, creates tables + seeds demo users)
 * 4) Login with seeded accounts (if you used init):
 *      manager: alice / Password123!
 *      employee: bob   / Password123!
 */

declare(strict_types=1);
session_start();

/* =========================
   CONFIG
   ========================= */
$dbHost = '127.0.0.1';
$dbName = 'employee_app';
$dbUser = 'root';
$dbPass = '';
$dsn = "mysql:host={$dbHost};dbname={$dbName};charset=utf8mb4";

/* =========================
   DB CONNECT
   ========================= */
function db(): PDO {
    static $pdo = null;
    if ($pdo instanceof PDO) return $pdo;

    global $dsn, $dbUser, $dbPass;
    $pdo = new PDO($dsn, $dbUser, $dbPass, [
        PDO::ATTR_ERRMODE => PDO::ERRMODE_EXCEPTION,
        PDO::ATTR_DEFAULT_FETCH_MODE => PDO::FETCH_ASSOC,
    ]);
    return $pdo;
}

/* =========================
   HELPERS
   ========================= */
function h(string $s): string { return htmlspecialchars($s, ENT_QUOTES | ENT_SUBSTITUTE, 'UTF-8'); }

function redirect(string $to): never {
    header("Location: {$to}");
    exit;
}

function current_user(): ?array {
    if (empty($_SESSION['uid'])) return null;

    $stmt = db()->prepare("SELECT id, username, role, employee_id, full_name, position FROM users WHERE id = ?");
    $stmt->execute([$_SESSION['uid']]);
    $u = $stmt->fetch();
    if (!$u) return null;
    return $u;
}

function require_login(): array {
    $u = current_user();
    if (!$u) redirect('?page=login');
    return $u;
}

function require_manager(array $u): void {
    if (($u['role'] ?? '') !== 'manager') {
        http_response_code(403);
        echo render_layout("Forbidden", "<p>403 - Forbidden (manager access required)</p>");
        exit;
    }
}

function is_role_valid(string $role): bool {
    return in_array($role, ['employee', 'manager'], true);
}

function csrf_token(): string {
    if (empty($_SESSION['csrf'])) {
        $_SESSION['csrf'] = bin2hex(random_bytes(32));
    }
    return $_SESSION['csrf'];
}

function csrf_verify(?string $token): bool {
    return is_string($token) && !empty($_SESSION['csrf']) && hash_equals($_SESSION['csrf'], $token);
}

/* =========================
   RENDERING
   ========================= */
function render_layout(string $title, string $bodyHtml, ?array $user = null): string {
    $nav = render_nav($user);
    $year = date('Y');

    return <<<HTML
<!doctype html>
<html lang="en">
<head>
  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width, initial-scale=1">
  <title>{$title}</title>
  <style>
    body { font-family: system-ui, -apple-system, Segoe UI, Roboto, Arial, sans-serif; margin: 0; background:#f6f7fb; color:#111; }
    header { background:#111827; color:#fff; padding: 14px 18px; }
    header .row { display:flex; justify-content:space-between; gap: 12px; align-items:center; flex-wrap: wrap; }
    header a { color:#fff; text-decoration:none; }
    main { max-width: 980px; margin: 20px auto; padding: 0 16px; }
    .card { background:#fff; border:1px solid #e5e7eb; border-radius: 10px; padding: 16px; box-shadow: 0 1px 2px rgba(0,0,0,.03); }
    .muted { color:#6b7280; }
    .nav a { margin-right: 10px; padding: 6px 10px; border-radius: 8px; display:inline-block; }
    .nav a:hover { background: rgba(255,255,255,.12); }
    .pill { display:inline-block; padding: 2px 8px; border-radius:999px; font-size: 12px; background:#eef2ff; color:#3730a3; }
    table { width:100%; border-collapse: collapse; margin-top: 10px; }
    th, td { text-align:left; padding: 10px 8px; border-bottom: 1px solid #e5e7eb; vertical-align: top; }
    th { font-size: 13px; color:#374151; }
    input, select { padding: 8px 10px; border-radius: 8px; border: 1px solid #d1d5db; width: 100%; box-sizing: border-box; }
    label { display:block; margin: 10px 0 6px; font-size: 13px; color:#374151; }
    button { padding: 9px 12px; border:0; border-radius: 9px; background:#2563eb; color:#fff; cursor:pointer; }
    button.secondary { background:#6b7280; }
    .row { display:flex; gap: 12px; flex-wrap: wrap; }
    .col { flex: 1 1 260px; }
    .msg { padding: 10px 12px; border-radius: 10px; margin-bottom: 12px; }
    .msg.ok { background:#ecfdf5; border:1px solid #a7f3d0; color:#065f46; }
    .msg.err { background:#fef2f2; border:1px solid #fecaca; color:#991b1b; }
    footer { max-width: 980px; margin: 18px auto 28px; padding: 0 16px; color:#6b7280; font-size: 13px; }
    code { background:#f3f4f6; padding: 2px 6px; border-radius: 6px; }
  </style>
</head>
<body>
<header>
  <div class="row">
    <div>
      <strong>Employee Management</strong>
      <span class="muted" style="color:#cbd5e1;">(single-file PHP)</span>
    </div>
    <div class="nav">{$nav}</div>
  </div>
</header>
<main>
  {$bodyHtml}
</main>
<footer>
  <div>© {$year} — Demo app. Use HTTPS + secure secrets in real deployments.</div>
</footer>
</body>
</html>
HTML;
}

function render_nav(?array $user): string {
    if (!$user) {
        return '<a href="?page=login">Login</a> <a href="?page=init">Init DB</a>';
    }

    $links = [];
    $links[] = '<a href="?page=profile">Profile</a>';
    if (($user['role'] ?? '') === 'manager') {
        $links[] = '<a href="?page=manage">Manage Employees</a>';
    }
    $links[] = '<a href="?page=logout">Logout</a>';

    $who = h($user['username'] . ' (' . $user['role'] . ')');
    return '<span class="pill">' . $who . '</span> ' . implode(' ', $links);
}

function msg_html(?string $ok = null, ?string $err = null): string {
    $out = '';
    if ($ok)  $out .= '<div class="msg ok">' . h($ok) . '</div>';
    if ($err) $out .= '<div class="msg err">' . h($err) . '</div>';
    return $out;
}

/* =========================
   ROUTES
   ========================= */
$page = $_GET['page'] ?? 'profile';

try {
    switch ($page) {
        case 'init':
            echo handle_init();
            break;

        case 'login':
            echo handle_login();
            break;

        case 'logout':
            $_SESSION = [];
            if (ini_get("session.use_cookies")) {
                $params = session_get_cookie_params();
                setcookie(session_name(), '', time() - 42000,
                    $params["path"], $params["domain"],
                    $params["secure"], $params["httponly"]
                );
            }
            session_destroy();
            redirect('?page=login');
            break;

        case 'profile':
            echo handle_profile();
            break;

        case 'manage':
            echo handle_manage();
            break;

        default:
            http_response_code(404);
            echo render_layout("Not found", "<div class='card'><p>404 - Page not found.</p></div>", current_user());
            break;
    }
} catch (Throwable $e) {
    http_response_code(500);
    $safe = h($e->getMessage());
    echo render_layout("Error", "<div class='card'><p><strong>Server error:</strong> {$safe}</p></div>", current_user());
}

/* =========================
   HANDLERS
   ========================= */
function handle_init(): string {
    // NOTE: In real apps, do not expose schema init publicly.
    $pdo = db();

    $pdo->exec("
        CREATE TABLE IF NOT EXISTS users (
            id INT AUTO_INCREMENT PRIMARY KEY,
            username VARCHAR(60) NOT NULL UNIQUE,
            password_hash VARCHAR(255) NOT NULL,
            role ENUM('employee','manager') NOT NULL DEFAULT 'employee',
            employee_id VARCHAR(30) NOT NULL,
            full_name VARCHAR(120) NOT NULL,
            position VARCHAR(120) NOT NULL,
            created_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP
        ) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4;
    ");

    // Seed demo users if not present
    $seed = [
        [
            'username' => 'alice',
            'password' => 'Password123!',
            'role' => 'manager',
            'employee_id' => 'M-1001',
            'full_name' => 'Alice Manager',
            'position' => 'Engineering Manager',
        ],
        [
            'username' => 'bob',
            'password' => 'Password123!',
            'role' => 'employee',
            'employee_id' => 'E-2001',
            'full_name' => 'Bob Employee',
            'position' => 'Software Engineer',
        ],
    ];

    $created = 0;
    $skipped = 0;

    $check = $pdo->prepare("SELECT id FROM users WHERE username = ?");
    $ins = $pdo->prepare("
        INSERT INTO users (username, password_hash, role, employee_id, full_name, position)
        VALUES (?, ?, ?, ?, ?, ?)
    ");

    foreach ($seed as $u) {
        $check->execute([$u['username']]);
        if ($check->fetch()) {
            $skipped++;
            continue;
        }
        $ins->execute([
            $u['username'],
            password_hash($u['password'], PASSWORD_DEFAULT),
            $u['role'],
            $u['employee_id'],
            $u['full_name'],
            $u['position'],
        ]);
        $created++;
    }

    $body = "<div class='card'>
        <h2>Database initialized</h2>
        " . msg_html("Tables ensured. Seeded: {$created}, already existed: {$skipped}.") . "
        <p>Try logging in at <a href='?page=login'>Login</a>.</p>
        <p class='muted'>Seeded accounts (if newly created): <code>alice / Password123!</code> (manager), <code>bob / Password123!</code> (employee)</p>
    </div>";

    return render_layout("Init DB", $body, current_user());
}

function handle_login(): string {
    $user = current_user();
    if ($user) redirect('?page=profile');

    $ok = null; $err = null;

    if ($_SERVER['REQUEST_METHOD'] === 'POST') {
        $username = trim((string)($_POST['username'] ?? ''));
        $password = (string)($_POST['password'] ?? '');
        $token = (string)($_POST['csrf'] ?? '');

        if (!csrf_verify($token)) {
            $err = "Invalid CSRF token. Refresh and try again.";
        } elseif ($username === '' || $password === '') {
            $err = "Username and password are required.";
        } else {
            $stmt = db()->prepare("SELECT id, username, password_hash, role FROM users WHERE username = ?");
            $stmt->execute([$username]);
            $row = $stmt->fetch();

            if (!$row || !password_verify($password, $row['password_hash'])) {
                $err = "Invalid username or password.";
            } else {
                session_regenerate_id(true);
                $_SESSION['uid'] = (int)$row['id'];
                $ok = "Logged in successfully.";
                redirect('?page=profile');
            }
        }
    }

    $token = h(csrf_token());
    $body = "<div class='card'>
        <h2>Login</h2>
        " . msg_html($ok, $err) . "
        <form method='post' action='?page=login'>
          <input type='hidden' name='csrf' value='{$token}'>
          <label>Username</label>
          <input name='username' autocomplete='username' required>
          <label>Password</label>
          <input name='password' type='password' autocomplete='current-password' required>
          <div style='margin-top:12px;'>
            <button type='submit'>Sign in</button>
            <a class='muted' style='margin-left:10px;' href='?page=init'>Init DB</a>
          </div>
        </form>
        <p class='muted' style='margin-top:12px;'>If you ran init, try <code>alice / Password123!</code> or <code>bob / Password123!</code>.</p>
    </div>";

    return render_layout("Login", $body, null);
}

function handle_profile(): string {
    $u = require_login();

    $body = "<div class='card'>
      <h2>Your Profile</h2>
      <table>
        <tr><th>Username</th><td>" . h($u['username']) . "</td></tr>
        <tr><th>Full name</th><td>" . h($u['full_name']) . "</td></tr>
        <tr><th>Employee ID</th><td>" . h($u['employee_id']) . "</td></tr>
        <tr><th>Position</th><td>" . h($u['position']) . "</td></tr>
        <tr><th>Role</th><td><span class='pill'>" . h($u['role']) . "</span></td></tr>
      </table>
      <p class='muted' style='margin-top:10px;'>Employees can only view their own profile. Managers can manage employee roles.</p>
    </div>";

    return render_layout("Profile", $body, $u);
}

function handle_manage(): string {
    $u = require_login();
    require_manager($u);

    $ok = null; $err = null;

    if ($_SERVER['REQUEST_METHOD'] === 'POST') {
        $token = (string)($_POST['csrf'] ?? '');
        if (!csrf_verify($token)) {
            $err = "Invalid CSRF token.";
        } else {
            $targetUserId = (int)($_POST['user_id'] ?? 0);
            $newRole = (string)($_POST['new_role'] ?? '');

            if ($targetUserId <= 0 || !is_role_valid($newRole)) {
                $err = "Invalid request.";
            } else {
                // Prevent manager from demoting themselves (optional safety)
                if ($targetUserId === (int)$u['id'] && $newRole !== 'manager') {
                    $err = "You cannot change your own role away from manager.";
                } else {
                    $stmt = db()->prepare("UPDATE users SET role = ? WHERE id = ?");
                    $stmt->execute([$newRole, $targetUserId]);
                    $ok = "Role updated.";
                }
            }
        }
    }

    $rows = db()->query("SELECT id, username, role, employee_id, full_name, position FROM users ORDER BY full_name ASC")->fetchAll();
    $token = h(csrf_token());

    $tableRows = '';
    foreach ($rows as $r) {
        $rid = (int)$r['id'];
        $role = (string)$r['role'];

        $options = '';
        foreach (['employee', 'manager'] as $opt) {
            $sel = ($opt === $role) ? 'selected' : '';
            $options .= "<option value='{$opt}' {$sel}>{$opt}</option>";
        }

        $tableRows .= "<tr>
          <td>" . h($r['full_name']) . "</td>
          <td>" . h($r['employee_id']) . "</td>
          <td>" . h($r['position']) . "</td>
          <td><span class='pill'>" . h($role) . "</span></td>
          <td>" . h($r['username']) . "</td>
          <td style='min-width:260px;'>
            <form method='post' action='?page=manage' class='row' style='gap:8px; align-items:center;'>
              <input type='hidden' name='csrf' value='{$token}'>
              <input type='hidden' name='user_id' value='{$rid}'>
              <div class='col' style='flex: 1 1 120px;'>
                <select name='new_role'>{$options}</select>
              </div>
              <div>
                <button type='submit'>Update</button>
              </div>
            </form>
          </td>
        </tr>";
    }

    $body = "<div class='card'>
      <h2>Manage Employees</h2>
      " . msg_html($ok, $err) . "
      <p class='muted'>As a manager, you can view all users and update their roles.</p>
      <table>
        <thead>
          <tr>
            <th>Name</th>
            <th>Employee ID</th>
            <th>Position</th>
            <th>Role</th>
            <th>Username</th>
            <th>Change role</th>
          </tr>
        </thead>
        <tbody>
          {$tableRows}
        </tbody>
      </table>
    </div>";

    return render_layout("Manage", $body, $u);
}
?>